pumpp/task/key.py

```diff
@@ -0,0 +1,438 @@
+            self.transition = None
+        else:
+            self.transition = transition_loop(len(self._classes), p_self)
+
+    def empty(self, duration):
+        '''Empty key annotations
+
+        Parameters
+        ----------
+        duration : number
+            The length (in seconds) of the empty annotation
+
+        Returns
+        -------
+        ann : jams.Annotation
+            A key annotation consisting of a single `N` observation.
+        '''
+        ann = super(KeyTagTransformer, self).empty(duration)
+
+        ann.append(time=0,
+                   duration=duration,
+                   value='N', confidence=0)
+
+        return ann
+
+    def vocabulary(self):
+        ''' Build the vocabulary for all key_mode strings
+
+        Returns
+        -------
+        labels : list
+            list of string labels.
+        '''
+        qualities = MODES + list(QUALITY.keys())
+        tonics = midi_to_note(list(range(12)), octave=False)
+        
+        labels = ['N']
+
+        for key_mode in product(tonics, qualities):
+            labels.append('{}:{}'.format(*key_mode))
+
+        return labels
+
+    def enharmonic(self, key_str):
+        '''Force the tonic spelling to fit our tonic list 
+        by spelling out of vocab keys enharmonically.
+
+        Parameters
+        ----------
+        key_str : str
+            The key_mode string in jams style.
+
+        Returns
+        -------
+        key_str : str
+            The key_mode string spelled enharmonically to fit our vocab.
+        '''
+        key_list = key_str.split(':')
+        # spell the tonic enharmonically if necessary
+        if key_list[0] != 'N':
+            key_list[0] = midi_to_note(note_to_midi(key_list[0]), octave=False)
+            if len(key_list) == 1:
+                key_list.append('major')
+
+        return ':'.join(key_list)
+
+    def transform_annotation(self, ann, duration):
+        '''Transform an annotation to key-tag encoding
+
+        Parameters
+        ----------
+        ann : jams.Annotation
+            The annotation to convert
+
+        duration : number > 0
+            The duration of the track
+
+        Returns
+        -------
+        data : dict
+            if self.sparse = True
+            data['keytag'] : np.ndarray, shape=(n, n_labels) or shape=(n,)
+                A time-varying binary encoding of the chords. 
+                The shape depends on self.sparse.
+        '''
+        intervals, values = ann.to_interval_values()
+
+        keys = []
+        for v in values:
+            keys.extend(self.encoder.transform([self.enharmonic(v)]))
+
+        dtype = self.fields[self.scope('keytag')].dtype
+
+        keys = np.asarray(keys)
+
+        if self.sparse:
+            keys = keys[:, np.newaxis]
+        
+        target = self.encode_intervals(duration, intervals, keys,
+                                       multi=False, dtype=dtype)
+
+        return {'keytag': target}
```

Comment:
I think we can just call this `tag`, since it will be scoped as `key_tag/tag` (or whatever the user names the transformer object).

Reply:
Naming things in code is hard for me hahaha. `tag` definitely works better.